p2p/p2p.md

```diff
@@ -0,0 +1,222 @@
+# P2P
+
+The p2p package mainly contains two services:
+
+1) Subscriber
+2) Exchange
+
+## Subscriber
+
+Subscriber is a service that manages gossip of headers among the nodes in the p2p network by using [libp2p][libp2p] and its [pubsub][pubsub] modules. The pubsub topic used for gossip (`/<networkID>/header-sub/v0.0.1`) is configurable based on `networkID` parameter used to initialize the subscriber service.
+
+The Subscriber implements the following interface:
+
+```
+// Subscriber encompasses the behavior necessary to
+// subscribe/unsubscribe from new Header events from the
+// network.
+type Subscriber[H Header[H]] interface {
+ // Subscribe creates long-living Subscription for validated Headers.
+ // Multiple Subscriptions can be created.
+ Subscribe() (Subscription[H], error)
+ // SetVerifier registers verification func for all Subscriptions.
+ // Registered func screens incoming headers
+ // before they are forwarded to Subscriptions.
+ // Only one func can be set.
+ SetVerifier(func(context.Context, H) error) error
+}
+```
+
+The `Subscribe()` method allows listening to any new headers that are published to the p2p network. The `SetVerifier()` method allows for setting a custom verifier that will be executed upon receiving any new headers from the p2p network. This is a very useful customization for the consumers of go-header library to pass any custom logic as part of the pubsub.
+
+## Exchange
+
+An exchange is a combination of:
+
+* Exchange: a client for requesting headers from the p2p network (outbound)
+* ExchangeServer: a p2p server for handling inbound header requests
+
+### Exchange Client
+
+Exchange defines a client for requesting headers from the p2p network. An exchange client is initialized using self [host.Host][host], a list of peers in the form of slice [peer.IDSlice][peer], and a [connection gater][gater] for blocking and allowing nodes. Optional parameters like `ChainID` and `NetworkID` can also be passed. The exchange client also maintains a list of trusted peers via a peer tracker.
+
+#### Peer Tracker
```

Comment:
I think it's worth adding that PeerTracker discovers peers until `len(connected)+len(disconnected)` will not reach the limit(for now it is 100)